manifests/init.pp

```diff
@@ -30,32 +40,53 @@
 #   *If* the $servers array above starts with '127.0.0.1' or '::1', then
 #   the system will set itself up as a caching nameserver unless this is set
 #   to false.
-#
+# @param sortlist
+#   Optional Array of address/netmask pairs that allow addresses returned by
+#   gethostbyname to be sorted.
+# @param extra_options
+#   Optional Array to put any options that may not be covered by the variables
+#   below. These will be appended to the options string.
 class resolv (
   Simplib::Netlist           $servers        = simplib::lookup('simp_options::dns::servers', { 'default_value' => ['127.0.0.1'] }),
   Simplib::Netlist           $search         = simplib::lookup('simp_options::dns::search', { 'default_value'  => [$facts['domain']] }),
-  Optional[Simplib::Netlist] $sortlist       = undef,
-  Optional[Array]            $extra_options  = undef,
-  String                     $resolv_domain  = $facts['domain'],
+  Resolv::Domain             $resolv_domain  = $facts['domain'],
   Boolean                    $debug          = false,
   Boolean                    $rotate         = true,
   Boolean                    $no_check_names = false,
   Boolean                    $inet6          = false,
-  Integer                    $ndots          = 1,
-  Integer                    $timeout        = 2,
-  Integer                    $attempts       = 2,
+  Integer[0,15]              $ndots          = 1,
+  Integer[0,30]              $timeout        = 2,
+  Integer[0,5]               $attempts       = 2,
   Boolean                    $named_server   = false,
   Boolean                    $named_autoconf = true,
-  Boolean                    $caching        = true
+  Boolean                    $caching        = true,
+  Optional[Resolv::Sortlist] $sortlist       = undef,
+  Optional[Array[String]]    $extra_options  = undef,
```

Comment:
There are only a handful of additional options listed in the man page.  If you are going to the trouble of validating everything else, why not validate them?  Why not have all of the missing options as parameters and note that $extra_options is deprecated and will be removed in the next release?

Reply:
Good questions, @lnemsick-simp.  

The short answer is that it's probably not worth doing―and possibly worth _not_ doing.

The `$extra_options` Array is a useful collection for less-common items from the "options" section of [resolv.conf](https://linux.die.net/man/5/resolv.conf).  It keeps the parameters focused on options that are relevant to the module, but still lets administrators manage relatively obscure or unforeseen options.  It's also trivial to understand and maintain.

Replacing this useful structure with a 1:1 mapping for each obscure option to both the class + template parameters seems disproportionate.  It makes the module's API more fragile and complex than it needs to be, without much benefit (other than completeness).

There are some obvious downsides, though:

* The extra parameters require more development (init.pp, resolv.epp, possibly new data types), documentation, tests, and subsequent maintenance. 
* This incurs time/effort/opportunity costs for a negligible return, and dilutes the overall signal:noise ratio around the repo with insignificant parameters.
* The entries for `$extra_options` are more likely to change over time or differ across OSes, but our definitions would be tightly coupled to EL6/EL7's resolv.conf.  Options that are only valid on certain OSes or OS versions will require more complex logic/module data to validate completely, which in turn would require additional releases.

This is pretty much business as usual for more relevant parameters, but―considering the relative insignificance of these options―it's disproportionate to spend effort to replace a simple and flexible solution with a more elaborate and brittle one.